coaster/annotations.py

```diff
@@ -0,0 +1,156 @@
+            elif hasattr(attr, '_coaster_annotations'):
+                data = attr._coaster_annotations
+            else:
+                data = None
+            if data is not None:
+                cls.__annotations_by_attr__.setdefault(name, []).extend(data)
+                for a in data:
+                    cls.__annotations__.setdefault(a, []).append(name)
+                processed.add(name)
+
+    annotations_configured.send(cls)
+
+
+@event.listens_for(mapper, 'after_configured')
+def __clear_cache():
+    for key in tuple(__cache__):
+        del __cache__[key]
+
+
+# --- Helpers -----------------------------------------------------------------
+
+def annotation_wrapper(annotation, doc=None):
+    """
+    Defines an annotation, which can be applied to attributes in a class
+    derived from :class:`AnnotationMixin`.
+    """
+    def decorator(attr):
+        __cache__.setdefault(attr, []).append(annotation)
+        # Also mark the annotation on the object itself. This will
+        # fail if the object has a restrictive __slots__, but it's
+        # required for some objects like Column because SQLAlchemy copies
+        # them in subclasses, changing their hash and making them
+        # undiscoverable via the cache.
+        try:
+            if not hasattr(attr, '_coaster_annotations'):
+                setattr(attr, '_coaster_annotations', [])
+            attr._coaster_annotations.append(annotation)
+        except AttributeError:
+            pass
+        return attr
+
+    decorator.__name__ = decorator.name = annotation
+    decorator.__doc__ = doc
+    return decorator
+
+
+# --- Annotations -------------------------------------------------------------
+
+immutable = annotation_wrapper('immutable', "Makes a column immutable once set")
+cached = annotation_wrapper('cached', "Marks the column's contents as a cached value from another source")
+
+
+# This code borrowed from https://stackoverflow.com/a/35352471/78903
+class ImmutableColumnError(AttributeError):
+    def __init__(self, class_name, column_name, old_value, new_value, message=None):
+        self.class_name = class_name
+        self.column_name = column_name
+        self.old_value = old_value
+        self.new_value = new_value
+
+        if message is None:
+            self.message = (
+                u"Cannot update column {column_name} on model {class_name} from {old_value} to {new_value}: "
+                u"column is immutable.".format(
+                    column_name=column_name, class_name=class_name, old_value=old_value, new_value=new_value))
+
+
+@annotations_configured.connect
+def __make_immutable(cls):
+    for attr in cls.__annotations__.get(immutable.name, []):
+        col = getattr(cls, attr)
+
+        @event.listens_for(col, 'set')
+        def immutable_column_set_listener(target, value, old_value, initiator):
+            if old_value != symbol('NEVER_SET') and old_value != symbol('NO_VALUE') and old_value != value:
```

Comment:
[SQLAlchemy's documentation suggests this use of NO_VALUE needs further testing](http://docs.sqlalchemy.org/en/latest/orm/internals.html#sqlalchemy.orm.state.AttributeState.loaded_value). Specifically, what happens if we load an instance from the database minus the column, _and then_ set that column? The old value will now be `NO_VALUE` and this validator will let it be updated, breaking immutability.